cpp/src/neighbors/cagra_c.cpp

```diff
@@ -263,6 +263,101 @@ void* _deserialize(cuvsResources_t res, const char* filename)
+  cuvs::neighbors::cagra::merge_params merge_params_cpp;
+  auto& out_idx_params = *params.output_index_params;
+
+  merge_params_cpp.output_index_params.metric =
+    static_cast<cuvs::distance::DistanceType>((int)out_idx_params.metric);
+  merge_params_cpp.output_index_params.intermediate_graph_degree =
+    out_idx_params.intermediate_graph_degree;
+  merge_params_cpp.output_index_params.graph_degree = out_idx_params.graph_degree;
+
+  switch (out_idx_params.build_algo) {
+    case cuvsCagraGraphBuildAlgo::AUTO_SELECT: break;
+
+    case cuvsCagraGraphBuildAlgo::IVF_PQ: {
+      int64_t total_size = 0;
+      auto first_idx_ptr =
+        reinterpret_cast<cuvs::neighbors::cagra::index<T, uint32_t>*>(indices[0]->addr);
+      int64_t dim = first_idx_ptr->dim();
+
+      for (size_t i = 0; i < num_indices; ++i) {
+        auto idx_ptr =
+          reinterpret_cast<cuvs::neighbors::cagra::index<T, uint32_t>*>(indices[i]->addr);
+        total_size += idx_ptr->size();
+      }
+
+      auto pq_params = cuvs::neighbors::cagra::graph_build_params::ivf_pq_params(
+        raft::matrix_extent<int64_t>(total_size, dim));
+
+      if (out_idx_params.graph_build_params) {
+        auto ivf_params = static_cast<cuvsIvfPqParams*>(out_idx_params.graph_build_params);
+        if (ivf_params->ivf_pq_build_params) {
+          auto bp                                         = ivf_params->ivf_pq_build_params;
+          pq_params.build_params.add_data_on_build        = bp->add_data_on_build;
+          pq_params.build_params.n_lists                  = bp->n_lists;
+          pq_params.build_params.kmeans_n_iters           = bp->kmeans_n_iters;
+          pq_params.build_params.kmeans_trainset_fraction = bp->kmeans_trainset_fraction;
+          pq_params.build_params.pq_bits                  = bp->pq_bits;
+          pq_params.build_params.pq_dim                   = bp->pq_dim;
+          pq_params.build_params.codebook_kind =
+            static_cast<cuvs::neighbors::ivf_pq::codebook_gen>(bp->codebook_kind);
+          pq_params.build_params.force_random_rotation = bp->force_random_rotation;
+          pq_params.build_params.conservative_memory_allocation =
+            bp->conservative_memory_allocation;
+          pq_params.build_params.max_train_points_per_pq_code = bp->max_train_points_per_pq_code;
+        }
+        if (ivf_params->ivf_pq_search_params) {
+          auto sp                                          = ivf_params->ivf_pq_search_params;
+          pq_params.search_params.n_probes                 = sp->n_probes;
+          pq_params.search_params.lut_dtype                = sp->lut_dtype;
+          pq_params.search_params.internal_distance_dtype  = sp->internal_distance_dtype;
+          pq_params.search_params.preferred_shmem_carveout = sp->preferred_shmem_carveout;
+        }
+        if (ivf_params->refinement_rate > 1.0f) {
+          pq_params.refinement_rate = ivf_params->refinement_rate;
+        }
+      }
```

Comment:
A bunch of this code seems cut and paste from the _build function - can we move into a function and share it so that we don't need to keep up to date in two spots?

Reply:
Done